library/std/src/sync/rwlock.rs

```diff
@@ -957,6 +961,29 @@ impl<'a, T: ?Sized> RwLockWriteGuard<'a, T> {
             None => Err(orig),
         }
     }
+
+    /// Downgrades a write-locked `RwLockWriteGuard` into a read-locked [`RwLockReadGuard`].
+    ///
+    /// Atomically changes the state of the [`RwLock`] from exclusive mode into shared mode.
+    ///
+    /// MORE DOCS COMING SOON.
+    #[unstable(feature = "rwlock_downgrade", issue = "128203")]
+    pub fn downgrade(s: Self) -> RwLockReadGuard<'a, T> {
+        let lock = s.lock;
+
+        // SAFETY: We take ownership of a write guard, so we must already have the `RwLock` in write
+        // mode, satisfying the `downgrade` contract.
+        unsafe { lock.inner.downgrade() };
+
+        // We don't want to call the destructor since that calls `write_unlock`.
+        forget(s);
+
+        // SAFETY: We have just successfully called `downgrade`, so we fulfill the safety contract.
+        unsafe {
+            RwLockReadGuard::new(lock)
+                .expect("We had the exclusive lock so we can't have panicked while holding it")
```

Comment:
Won't this fail if the lock was already poisoned when the `write` call acquired it? Consider someone doing
```rust
let write = lock.write().unwrap_or_else(PoisonError::into_inner);
```

Reply:
Ah I was not aware you were allowed to still acquire the write lock this way even if it was poisoned.

This means that the API is going to have to change to return a `LockResult`

Reply:
Are you saying that the downgrade method should clear the poison if it occurred?

Reply:
No, it should just leave it untouched and ignore it.